jlm/llvm/opt/alias-analyses/OnlineCycleDetection.hpp

```diff
@@ -0,0 +1,368 @@
+/*
+ * Copyright 2024 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_LLVM_OPT_ALIAS_ANALYSES_ONLINECYCLEDETECTION_HPP
+#define JLM_LLVM_OPT_ALIAS_ANALYSES_ONLINECYCLEDETECTION_HPP
+
+#include <jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp>
+#include <jlm/util/TarjanScc.hpp>
+
+#include <stack>
+#include <vector>
+
+namespace jlm::llvm::aa
+{
+
+/**
+ * Online Cycle Detection maintains a topological ordering of all unification roots, with respect
+ * to superset edges. If this is not possible, there must be a cycle, which gets eliminated.
+ * Online Cycle Detection can not be combined with other cycle detection schemes,
+ * but that is OK, since OCD finds every cycle.
+ * @tparam GetSuccessorsFunctor
+ * @tparam UnifyPointerObjectsFunctor
+ */
+template<typename GetSuccessorsFunctor, typename UnifyPointerObjectsFunctor>
+class OnlineCycleDetector
+{
+
+public:
+  OnlineCycleDetector(
+      PointerObjectSet & set,
+      const GetSuccessorsFunctor & GetSuccessors,
+      const UnifyPointerObjectsFunctor & UnifyPointerObjects)
+      : Set_(set),
+        GetSuccessors_(GetSuccessors),
+        UnifyPointerObjects_(UnifyPointerObjects)
+  {}
+
+  /**
+   * A assigns each unification root a position in a topological ordering.
+   * If any cycles are detected, they are unified, but nothing is placed on the worklist.
+   * This function assumes that GetSuccessors returns only unification roots.
+   */
+  void
+  InitializeTopologicalOrdering()
+  {
+    // Initialize Online Cycle Detection using a full topological sort of the nodes
+    ObjectToTopoOrder_.resize(Set_.NumPointerObjects(), -1);
+
+    // At this point, all subsetEdges are between unification roots only
+    // Use TarjanSCC to find our starting topological ordering
+    std::vector<size_t> sccIndex;
+    std::vector<size_t> topologicalOrder;
+
+    util::FindStronglyConnectedComponents(
+        Set_.NumPointerObjects(),
+        GetSuccessors_,
+        sccIndex,
+        topologicalOrder);
+
+    // Go through the topological ordering and add all unification roots to OCD_ObjectToTopoOrder
+    // Also, if we find any new SCCs while doing this, perform unification right now
+    for (size_t i = 0; i < topologicalOrder.size(); i++)
+    {
+      PointerObjectIndex node = topologicalOrder[i];
+      if (!Set_.IsUnificationRoot(node))
+        continue; // We only care about unification roots
+
+      // If this root belongs to the same scc as the next root, they should be unified
+      if (i + 1 < topologicalOrder.size() && sccIndex[node] == sccIndex[topologicalOrder[i + 1]])
+      {
+        // We know that the SCC consists of only roots, since GetSuccessors_ returns only roots
+        JLM_ASSERT(Set_.IsUnificationRoot(topologicalOrder[i + 1]));
+        // Make the next object the root, to keep the unification going
+        topologicalOrder[i + 1] = UnifyPointerObjects_(node, topologicalOrder[i + 1]);
+      }
```

Comment:
Introduce a variable `nextIndex = i + 1` (or a more suitable name) and use this one instead of the `i +1`s all of over the place.